File: thread-pool/src/task.go
package threadPool

// Example implementation of Callable interface
type Task struct {
	Func        func() (interface{}, error)
	Result      interface{}
	Err         error
	IsCompleted bool
	syncChan    chan struct{}
}

func NewTask(f func() (interface{}, error)) *Task {
	return &Task{Func: f, syncChan: make(chan struct{})}
}

func (t *Task) Call() {
	t.Result, t.Err = t.Func()
	t.IsCompleted = true
	close(t.syncChan)
}

func (t *Task) ContinueWith(f func (interface{}) (interface{}, error)) Callable {
	if !t.IsCompleted { <-t.syncChan }

	return NewTask(func() (interface{}, error) {
		return f(t.Result)
	})
}

File: Task2/Task2/Task.cs
using System;

namespace Task2
{
    public interface ITask
    {
        bool IsCompleted();
        bool CanBeExecuted();
        void Execute();
    }

    public interface ITask<T> : ITask
    {
        T Result();
        ITask<S> ContinueWith<S>(Func<T, S> function);
    }
}
